src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpointManagerBase.cs

```diff
@@ -96,15 +100,199 @@ protected ServiceEndpoint[] GetValuableEndpoints(IEnumerable<ServiceEndpoint> en
             return groupedEndpoints.ToArray();
         }
 
+        protected async Task AddServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints, TimeSpan timeout)
+        {
+            if (endpoints.Count > 0)
+            {
+                try
+                {
+                    var hubEndpoints = CreateHubServiceEndpoints(endpoints, true);
+
+                    await Task.WhenAll(hubEndpoints.Select(e => AddHubServiceEndpointAsync(e, timeout)));
+
+                    // TODO: update local store for negotiation
+                }
+                catch
+                {
+                    Log.FailedAddingEndpoints(_logger);
+                }
+            }
+        }
+
+        protected async Task RemoveServiceEndpointsAsync(IReadOnlyList<ServiceEndpoint> endpoints, TimeSpan timeout)
+        {
+            if (endpoints.Count > 0)
+            {
+                try
+                {
+                    var hubEndpoints = CreateHubServiceEndpoints(endpoints, true);
+
+                    // TODO: update local store for negotiation
+
+                    await Task.WhenAll(hubEndpoints.Select(e => RemoveHubServiceEndpointAsync(e, timeout)));
+                }
+                catch
+                {
+                    Log.FailedRemovingEndpoints(_logger);
+                }
+            }
+        }
+
+        protected Task RenameSerivceEndpoints(IReadOnlyList<ServiceEndpoint> endpoints)
+        {
+            if (endpoints.Count > 0)
+            {
+                try
+                {
+                    var hubEndpoints = CreateHubServiceEndpoints(endpoints, false);
+
+                    // TODO: update local store for negotiation
+
+                    return Task.WhenAll(hubEndpoints.Select(e => RenameHubServiceEndpoint(e)));
+                }
+                catch
+                {
+                    Log.FailedRenamingEndpoint(_logger);
+                }
+            }
+            return Task.CompletedTask;
+        }
+
+        private HubServiceEndpoint CreateHubServiceEndpoint(string hub, ServiceEndpoint endpoint, bool needScaleTcs = false)
+        {
+            var provider = GetEndpointProvider(endpoint);
+
+            TaskCompletionSource<bool> scaleTcs = null;
+            if (needScaleTcs)
```

Comment:
this makes things complicated, so the consumers need to check null everytime when call it? can we unify the conditions?